drivers/vmwarevsphere/cloudinit.go

```diff
@@ -0,0 +1,272 @@
+						Value: encoded,
+					})
+					opts = append(opts, &types.OptionValue{
+						Key:   "guestinfo.cloud-init.data.encoding",
+						Value: "base64",
+					})
+				}
+			}
+		}
+	}
+
+	return d.applyOpts(vm, opts)
+}
+
+func (d Driver) uploadCloudInitIso(vm *object.VirtualMachine, dc *object.Datacenter, dss *object.Datastore) error {
+	path, err := d.getFolder(vm)
+	if err != nil {
+		return err
+	}
+
+	dsurl, err := dss.URL(d.getCtx(), dc, filepath.Join(path, isoName))
+	if err != nil {
+		return err
+	}
+
+	p := soap.DefaultUpload
+	c, err := d.getSoapClient()
+	if err != nil {
+		return err
+	}
+
+	if err = c.Client.UploadFile(d.getCtx(), d.ResolveStorePath(filepath.Join(isoDir, isoName)), dsurl, &p); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d Driver) removeCloudInitIso(vm *object.VirtualMachine, dc *object.Datacenter, dss *object.Datastore) error {
+	c, err := d.getSoapClient()
+	if err != nil {
+		return err
+	}
+
+	path, err := d.getFolder(vm)
+	if err != nil {
+		return err
+	}
+
+	m := object.NewFileManager(c.Client)
+	task, err := m.DeleteDatastoreFile(d.getCtx(), dss.Path(filepath.Join(path, isoName)), dc)
+	if err != nil {
+		return err
+	}
+
+	if err = task.Wait(d.getCtx()); err != nil {
+		if types.IsFileNotFound(err) {
+			// already deleted, ignore error
+			return nil
+		}
+
+		return err
+	}
+
+	return nil
+}
+
+func (d Driver) createCloudInitIso() error {
+	if _, err := os.Stat(d.CloudInit); err != nil {
+		return err
+	}
+
+	perm := os.FileMode(0700)
+	isoDir := d.ResolveStorePath(isoDir)
+	dataDir := filepath.Join(isoDir, "data")
+	userdata := filepath.Join(dataDir, "user-data")
+	metadata := filepath.Join(dataDir, "meta-data")
+
+	err := os.MkdirAll(dataDir, perm)
+	if err != nil {
+		return err
+	}
+
+	ud, err := ioutil.ReadFile(d.CloudInit)
+	if err != nil {
+		return err
+	}
+
+	cf := make(map[interface{}]interface{})
+	err = yaml.Unmarshal([]byte(ud), &cf)
+	if err != nil {
+		return err
+	}
+
+	pubKey, err := ioutil.ReadFile(d.publicSSHKeyPath())
+	if err != nil {
+		return err
+	}
+
+	newUser := map[interface{}]interface{}{
+		"name":        "docker",
+		"sudo":        "ALL=(ALL) NOPASSWD:ALL",
+		"lock_passwd": "true",
+		"ssh_authorized_keys": []string{
+			string(pubKey),
+		},
+	}
+
+	if val, ok := cf["users"]; ok {
+		u := val.([]interface{})
+		users := make([]interface{}, len(u)+1)
+		for k, v := range u {
+			users[k] = v
+		}
+		users[len(users)-1] = newUser
+		cf["users"] = users
+	} else {
+		users := make([]interface{}, 1)
+		users[0] = newUser
+		cf["users"] = users
+	}
+
+	cfBody, err := yaml.Marshal(cf)
+	if err != nil {
+		return err
+	}
+
+	writeYaml := fmt.Sprintf("#cloud-config\n%s", cfBody)
+	ioutil.WriteFile(userdata, []byte(writeYaml), perm)
+
+	md := []byte(fmt.Sprintf("#local-hostname: %s\n", d.MachineName))
+	if err = ioutil.WriteFile(metadata, md, perm); err != nil {
+		return err
+	}
+
+	//making iso
+	blocksize := int64(2048)
```

Comment:
Feels weird to put all this stuff inline rather than configs, guessing it wouldn't ever need to change.